src/phase0/state_transition/epoch_processing.rs

```diff
@@ -444,6 +527,430 @@ pub fn process_participation_record_updates<
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    attestations: impl Iterator<Item = &'a PendingAttestation<MAX_VALIDATORS_PER_COMMITTEE>>,
+    context: &Context,
+) -> Result<(Vec<Gwei>, Vec<Gwei>), Error> {
+    // Helper with shared logic for use by get source, target, and head deltas functions
+    let validator_count = state.validators.len();
+    let mut rewards = vec![0; validator_count];
+    let mut penalties = vec![0; validator_count];
+    let total_balance = get_total_active_balance(state, context)?;
+    let unslashed_attesting_indices =
+        get_unslashed_attesting_indices(state, attestations, context)?;
+    let attesting_balance = get_total_balance(state, &unslashed_attesting_indices, context)?;
+    let increment = context.effective_balance_increment; // Factored out from balance totals to avoid uint64 overflow
+    for i in get_eligible_validator_indices(state, context) {
+        if unslashed_attesting_indices.contains(&i) {
+            if is_in_inactivity_leak(state, context) {
+                // Since full base reward will be canceled out by inactivity penalty deltas,
+                // optimal participation receives full base reward compensation here.
+                rewards[i] += get_base_reward(state, i, context)?;
+            } else {
+                let reward_numerator =
+                    get_base_reward(state, i, context)? * (attesting_balance / increment);
+                rewards[i] += reward_numerator / (total_balance / increment);
+            }
+        } else {
+            penalties[i] += get_base_reward(state, i, context)?;
+        }
+    }
+    Ok((rewards, penalties))
+}
+
+pub fn get_source_deltas<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> Result<(Vec<Gwei>, Vec<Gwei>), Error> {
+    // Return attester micro-rewards/penalties for source-vote for each validator.
+    let previous_epoch = get_previous_epoch(state, context);
+    let matching_source_attestations =
+        get_matching_source_attestations(state, previous_epoch, context)?.iter();
+    get_attestation_component_deltas(state, matching_source_attestations, context)
```

Comment:
a bit minor but you did this the other way in other places and i have a preference for that as the thing that is bound by a name is (more) concrete data, rather than something like an "Iter"
```suggestion
    let matching_source_attestations =
        get_matching_source_attestations(state, previous_epoch, context)?;
    get_attestation_component_deltas(state, matching_source_attestations.into_iter(), context)
```

and i think you can do `into_iter` over just `iter` here

Reply:
Good catch -- I can definitely at least put the `iter()` within the fn call. However, using `matching_source_attestations.into_iter()` brings a couple of errors:
1. `get_attestation_component_deltas` expects a reference, but `into_iter` makes it a value -- i.e., I get an error `expected reference &PendingAttestation, found struct PendingAttestation`
2. Thus, if I use `&matching_source_attestations.into_iter()`, then I run into iterator issues: `&IntoIter<PendingAttestation>` is not an iterator

Fwiw, if I alter the return type on `get_matching_source_attestations` to a simple `Vec`, `into_iter` (and everything else still) works -- so it seems like it's an `ssz_rs`-related aspect?

Generally speaking, is `into_iter` desired since it's more flexible than `iter`? E.g., yields `T`, `&T` or `&mut T` instead of only` &T`

Reply:
> Generally speaking, is into_iter desired since it's more flexible than iter? E.g., yields T, &T or &mut T instead of only &T

that is perhaps one way of looking at it -- `into_iter` yields an iterator by value, so that you get T instead of `iter` over &T or `iter_mut` over &mut T; i suppose there are situations where the compiler will insert borrows for you to match the call site if you use `into_iter`

my preference is to use "by value" over "by ref" where we can, but it is just an aesthetic thing and may have no bearing as the compiler may turn around and do crazy stuff w/ the code based on the context... the aesthetics here are to follow "sources and sinks" -- don't leave a thing lying around if you can just send the whole thing into a (e.g. trailing) function call

Reply:
and yeah i think i mislead w/ the `into_iter` call, just `iter` is fine here